### PR TITLE
Major Refactor

### PR DESCRIPTION
**Pull request by martintreurnicht**
We've forked this library to be used in our own projects, I figured these might be some good quality of life changes to be sent upstream:

* Switch to Typescript and using function component and hooks instead
* Convert iOS code to Swift instead (enables us to drop internal copy of `com.redmadrobot` library and use public release instead)
* Use latest versions of `com.redmadrobot` libraries

### CODE REVIEW

File: android/src/main/java/com/RNTextInputMask/RNTextInputMaskModule.java [hunk 1]
```diff
@@ -40,9 +40,9 @@ public void mask(final String maskString,
       final Mask.Result result = mask.apply(
           new CaretString(
               input,
-              input.length()
-          ),
-          true
+              input.length(),
+              new CaretString.CaretGravity.FORWARD(false)
```
Comment (martintreurnicht) [inline android/src/main/java/com/RNTextInputMask/RNTextInputMaskModule.java:44]: Api has changed, even though i've switch autocomplete off here it still autocompletes forward. If we switch this to true, then autocomplete causes problems when trying to clear out entered value

File: index.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      setValue(defaultValue)
+    }
+    const nodeId = findNodeHandle(input.current)
+    if (inputMask && nodeId) {
+      setMask(nodeId, inputMask)
+    }
+  }, [maskDefaultValue, inputMask, defaultValue])
+
+  useImperativeHandle(ref, () => ({
+    focus: () => {
```
Comment (martintreurnicht) [inline index.tsx:26]: exposing focus handlers, something we use and i figured it doesn't hurt and might be useful for others

**Comment by ivanzotov** [comment]
@martintreurnicht I see you did a lot of work here. Would you like to be a maintainer of the repo? Because I don't have much time to support it.

**Comment by martintreurnicht** [comment]
> @martintreurnicht I see you did a lot of work here. Would you like to be a maintainer of the repo? Because I don't have much time to support it.

Hi @ivanzotov, yes wouldn't mind assisting. We would anyway need to maintain a fork at minimum, would be nice to share the improvements with everyone

**Comment by ivanzotov** [comment]
@martintreurnicht That's great. I've sent you an invite to be an admin of the organisation.

**Comment by ivanzotov** [comment]
You have access to the npmjs now. I happy this project has a new owner now. Thank you!

**Comment by martintreurnicht** [comment]
> You have access to the npmjs as well. I happy this project has a new owner now. Thank you!

Thanks for the invite!

**Comment by betoharres** [comment]
Hey guys, I just found this PR and I was actually working on the same thing yesterday(I should have looked into this first, duh!).
I don't think I got the same level of code as @martintreurnicht but I'd like to share my simple code:

**Comment by betoharres** [comment]
Hey guys, I just found this PR and I was actually working on the same thing yesterday(I should have looked into this first, duh!).
I don't think I got the same level of code as @martintreurnicht but I'd like to share my simple code:
https://github.com/betoharres/react-native-text-input-mask

**Comment by betoharres** [comment]
Hey guys, I just found this PR and I was actually working on the same thing yesterday(I should have looked into this first, duh!).
I don't think I got the same level of code as @martintreurnicht but I'd like to share my simple code:
https://github.com/betoharres/react-native-text-input-mask

I'm not using typescript though! Just simple JSX. 
I also noticed that the android native dependency is outdated, and I already tried to bump the version number on `build.gradle` file but for some reason it can't download during build(not published on jcenter maybe?)

Also, I was testing this changes on my current project and I noticed that `React.forwardRef` doesn't work well with `react-native-elements` `InputComponent` prop. For some reason if you use `forwardRef` along with their `Input` component, it will make your component unmount between re-renders, thus loosing focus on every character you enter - similar to [this](https://github.com/formium/formik/issues/882#issuecomment-419589385). I spent the entire day today trying to figure out how to make it work but I gave up. It does work fine if you only use the `TextInputMask` component though.

Anyway, those were my cents, I'm happy to help if you guys need!

**Comment by martintreurnicht** [comment]
> Hey guys, I just found this PR and I was actually working on the same thing yesterday(I should have looked into this first, duh!).
> I don't think I got the same level of code as @martintreurnicht but I'd like to share my simple code:
> https://github.com/betoharres/react-native-text-input-mask
> 
> I'm not using typescript though! Just simple JSX.
> I also noticed that the android native dependency is outdated, and I already tried to bump the version number on `build.gradle` file but for some reason it can't download during build(not published on jcenter maybe?)
> 
> Also, I was testing this changes on my current project and I noticed that `React.forwardRef` doesn't work well with `react-native-elements` `InputComponent` prop. For some reason if you use `forwardRef` along with their `InputComponent`, it will make your component unmount between re-renders, thus loosing focus on every character you enter - similar to [this](https://github.com/formium/formik/issues/882#issuecomment-419589385). I spent the entire day today trying to figure out how to make it work but I gave up. It does work fine if you only use the `TextInputMask` component though.
> 
> Anyway, those were my cents, I'm happy to help if you guys need!

@betoharres i'll merge these changes soon, I've found some new issues that i'm in the process of resolving. Interesting comment about forwardRef. I'm not sure why it would cause it to unmount, should not be the case ( I haven't noticed any issues on my side). I'll look into it, but If you know what the issue is, PRs are welcome :-)

**Comment by betoharres** [comment]
@martintreurnicht sorry I think you thought I was talking about your changes, but the problems I mentioned were actually from my changes using `React.forwardRef`; I'll test your changes as soon as you release them :)

**Comment by martintreurnicht** [comment]
> @martintreurnicht sorry I think you thought I was talking about your changes, but the problems I mentioned were actually from my changes using `React.forwardRef`; I'll test your changes as soon as you release them :)

@betoharres Sorry for the delay with merging this PR. I've run into some other issues that i'm trying to fix all in one go. Hopefully i'll merge this over the weekend.

**Comment by betoharres** [comment]
> > @martintreurnicht sorry I think you thought I was talking about your changes, but the problems I mentioned were actually from my changes using `React.forwardRef`; I'll test your changes as soon as you release them :)
> 
> @betoharres Sorry for the delay with merging this PR. I've run into some other issues that i'm trying to fix all in one go. Hopefully i'll merge this over the weekend.

@martintreurnicht no worries mate, take your time :)

**Comment by SrBrahma** [comment]
Sounds promising!

**Comment by SrBrahma** [comment]
Huge thanks for the update! Will try the 3.0.0 soon. Does it need `use_frameworks!` in ios setup?

**Comment by martintreurnicht** [comment]
@SrBrahma Yes, not a requirement from this project but from upstream dependency https://github.com/RedMadRobot/input-mask-ios. I will update README to indicate `use_frameworks!` requirement

**Comment by SrBrahma** [comment]
I see. Thanks!

**Comment by betoharres** [comment]
wow awesome work, thanks a lot @martintreurnicht! I'll also check the new version soon; new code looks great! :tada:

**Pull request merged**